js-jest/src/gilded_rose.js

```diff
@@ -9,201 +9,141 @@ class Item {
+  constructor(sellIn, quality){
+    super(sellIn, quality)
+    this.name = 'Sulfuras, Hand of Ragnaros'
   }
 
-  _updateLegendaryItem(index){
-    this.items[index].sellIn = 0
+  updateItem(){
+    this.sellIn = 0
 
-    if(this.items[index].quality > 50){
-      this.items[index].quality = 50
+    if(this.quality > 50){
+      this.quality = 50
     }
+  }
+}
 
-    return this.items[index]
+class ConcertTickets extends Item {
+  constructor(sellIn, quality){
+    super(sellIn, quality)
+    this.name = 'Backstage passes to a TAFKAL80ETC concert'
   }
 
-  _updateConcertTickets(index){
-    const sellBy = this.items[index].sellIn
+  updateItem(){
+    const sellBy = this.sellIn
+
     switch(true){
       case (sellBy <= 0): 
-        this.items[index].quality = 0
+        this.quality = 0
         break
 
       case (sellBy > 0 && sellBy <= 5):
-        this.items[index].quality += 3
-        if (this.items[index].quality > 50){
-          this.items[index].quality = 50
+        this.quality += 3
+        if (this.quality > 50){
+          this.quality = 50
         }
         break
 
       case (sellBy > 5 && sellBy <= 10):
-        this.items[index].quality +=2
-        if (this.items[index].quality > 50){
-          this.items[index].quality = 50
+        this.quality +=2
+        if (this.quality > 50){
+          this.quality = 50
         }
         break
 
       case (sellBy > 10):
-        this.items[index].quality += 1
-        if (this.items[index].quality > 50){
-          this.items[index].quality = 50
+        this.quality += 1
+        if (this.quality > 50){
+          this.quality = 50
         }
         break
     }
   }
+}
 
-  updateQuality() {
-    for (let i = 0; i < this.items.length; i++) {
-      const itemName = this.items[i].name
-
-      switch(true){
-        case (itemName === 'Backstage passes to a TAFKAL80ETC concert'):
-          this._updateConcertTickets(i)
-          break
-
-        case (itemName === "Sulfuras, Hand of Ragnaros"):
-          this._updateLegendaryItem(i)
-          break
+class ConjuredItem extends Item {
+  constructor(sellIn, quality){
+    super(sellIn, quality)
+    this.name = "Conjured Item"
+  }
 
-        case (itemName === 'Aged Brie'):
-          this._updateAgedBrie(i)
-          break
+  updateItem(){
+    if (this.sellIn >0){
+      this.quality -= 2
 
-        case (itemName === "Conjured Item"):
-          this._updateConjuredItem(i)
-          break
+      if(this.quality < 0){
+        this.quality = 0
+      }
+    } else{
+      this.items[index].quality -= 4
 
-        default:
-          this._updateNormalItem(i)
-          break
+      if(this.items[index].quality < 0){
+        this.items[index].quality = 0
       }
     }
+  }
+}
 
-    return this.items;
+class Shop {
+  constructor(items=[]){
+    this.items = items;
+  }
+
+  updateQuality() {
+    for (let i = 0; i < this.items.length; i++) {
+      this.items[i].updateItem()
+    }
```

Comment:
This is 100% valid - but now might be a good opportunity to check out some other modern JS iterating methods:
[for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)
[forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
[good ol' map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)

One nice thing about them is that they can sometimes help make things a little more human-readable. For example:
```suggestion
    for (const item of this.items) {
      item.updateItem()
    }
```

or

```suggestion
    this.items.forEach(item => item.updateItem())
```

Again – absolutely nothing wrong with the way you did it though!

Reply:
Features implemented in newest update